File: src/2025-03-04-03-28-40--deepseek-chat-0304/java/exercises/practice/connect/src/main/java/Connect.java
class Connect {
    private String[] board;
    private int rows;
    private int cols;

    public Connect(String[] board) {
        this.board = board;
        this.rows = board.length;
        // Calculate columns by counting non-space characters in first row
        this.cols = 0;
        if (rows > 0) {
            for (int i = 0; i < board[0].length(); i++) {
                if (board[0].charAt(i) != ' ') {
                    cols++;
                }
            }
        }
    }

    public Winner computeWinner() {
        // Check if O has connected top to bottom
        if (hasPath('O', true)) {
            return Winner.PLAYER_O;
        }
        
        // Check if X has connected left to right
        if (hasPath('X', false)) {
            return Winner.PLAYER_X;
        }
        
        return Winner.NONE;
    }

    private boolean hasPath(char player, boolean vertical) {
        boolean[][] visited = new boolean[rows][cols];
        
        // Initialize starting positions
        for (int i = 0; i < (vertical ? cols : rows); i++) {
            int row = vertical ? 0 : i;
            int col = vertical ? i : 0;
            
            if (getCell(row, col) == player && 
                dfs(row, col, player, visited, vertical)) {
                return true;
            }
        }
        return false;
    }

    private boolean dfs(int row, int col, char player, boolean[][] visited, boolean vertical) {
        // Check if we've reached the opposite side
        if (vertical && row == rows - 1) return true;
        if (!vertical && col == cols - 1) return true;

        visited[row][col] = true;

        // Hex neighbors: right, left, down-right, down-left, up-right, up-left
        int[][] directions = {
            {0, 1}, {0, -1},
            {1, 0}, {1, 1},
            {-1, 0}, {-1, -1}
        };

        for (int[] dir : directions) {
            int newRow = row + dir[0];
            int newCol = col + dir[1];

            if (isValid(newRow, newCol) && 
                !visited[newRow][newCol] && 
                getCell(newRow, newCol) == player) {
                if (dfs(newRow, newCol, player, visited, vertical)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean isValid(int row, int col) {
        return row >= 0 && row < rows && col >= 0 && col < cols;
    }

    private char getCell(int row, int col) {
        // Skip spaces in the board representation
        String line = board[row];
        int actualCol = 0;
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c != ' ') {
                if (actualCol == col) {
                    return c;
                }
                actualCol++;
            }
        }
        return ' '; // Invalid cell
    }

    private boolean isValid(int row, int col) {
        if (row < 0 || row >= rows || col < 0) return false;
        // Calculate actual columns considering spaces
        if (rows == 0) return false;
        String line = board[row];
        int actualCols = 0;
        for (int i = 0; i < line.length(); i++) {
            if (line.charAt(i) != ' ') {
                actualCols++;
            }
        }
        return col < actualCols;
    }
}

enum Winner {
    PLAYER_X,
    PLAYER_O,
    NONE
}

File: src/2025-03-04-03-28-40--deepseek-chat-0304/javascript/exercises/practice/connect/connect.js
export class Board {
  constructor(board) {
    this.board = board.map(row => row.replace(/ /g, ''));
    this.height = this.board.length;
    this.width = this.height > 0 ? this.board[0].length : 0;
  }

  winner() {
    // Check for O's vertical connection
    if (this.checkConnection('O', 'vertical')) {
      return 'O';
    }
    // Check for X's horizontal connection
    if (this.checkConnection('X', 'horizontal')) {
      return 'X';
    }
    return '';
  }

  checkConnection(player, direction) {
    const visited = new Set();
    const toVisit = [];
    
    // Initialize starting positions
    if (direction === 'vertical') {
      for (let x = 0; x < this.width; x++) {
        if (this.getCell(x, 0) === player) {
          toVisit.push([x, 0]);
        }
      }
    } else {
      for (let y = 0; y < this.height; y++) {
        if (this.getCell(0, y) === player) {
          toVisit.push([0, y]);
        }
      }
    }

    // Perform BFS
    while (toVisit.length > 0) {
      const [x, y] = toVisit.pop();
      const key = `${x},${y}`;
      
      if (visited.has(key)) continue;
      visited.add(key);

      // Check if reached the opposite side
      if (direction === 'vertical' && y === this.height - 1) {
        return true;
      }
      if (direction === 'horizontal' && x === this.width - 1) {
        return true;
      }

      // Add neighbors
      for (const [dx, dy] of this.getNeighbors(x, y)) {
        if (this.getCell(dx, dy) === player) {
          toVisit.push([dx, dy]);
        }
      }
    }

    return false;
  }

  getCell(x, y) {
    if (x < 0 || x >= this.width || y < 0 || y >= this.height) {
      return null;
    }
    return this.board[y][x];
  }

  getNeighbors(x, y) {
    return [
      [x - 1, y],    // left
      [x + 1, y],    // right
      [x, y - 1],    // top-left
      [x + 1, y - 1],// top-right
      [x, y + 1],    // bottom-left
      [x - 1, y + 1] // bottom-right
    ].filter(([nx, ny]) => this.getCell(nx, ny) !== null);
  }
}
